### PR TITLE
[MM-55223] add basic dump mechansim

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
WIP: Create a tsdb dump and write zip file to the remote storage

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-55223

### CODE REVIEW

**Comment by isacikgoz** [comment]
The interval selection screen is something like this:


https://github.com/mattermost/mattermost-plugin-metrics/assets/2153367/cf71680f-3fae-4928-bdb7-73821c1bd68a

File: server/api.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	root.Use(handler.authorized)
+
+	root.HandleFunc("/download", handler.downloadDumpHandler)
+	handler.router = root
+
+	return handler
+}
+
+func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	r.Body = http.MaxBytesReader(w, r.Body, MaxRequestSize)
```
[the remark this answers is not in the document]
Comment (isacikgoz) [reply server/api.go:35]: Not really, it's just a extra measure. In fact I don't think this plugin even needs body limit as we are only doing a `GET` request.

File: server/dump.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+			// and to not pollute the logs, we simply reducing the log level here.
+			p.API.LogDebug("unable to read meta file", "err", rErr)
+			continue
+		}
+
+		metaMax := time.UnixMilli(meta.MaxTime)
+		if metaMax.Before(max) && metaMax.After(min) {
+			p.API.LogInfo("Fetching block from the filestore", "ulid", meta.ULID, "Max Time", max.String())
+
+			err = readFromFileStore(dumpDir, b, p.fileBackend)
```
[the remark this answers is not in the document]
Comment (isacikgoz) [reply server/dump.go:40]: Makes sense.

File: server/dump.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	blocks, err := p.fileBackend.ListDirectory(remoteStorageDir)
+	if err != nil {
+		return "", err
+	} else if len(blocks) == 0 {
+		return "", errors.New("no blocks in the remote sotrage")
+	}
+
+	zipFileNameRemote := filepath.Join(pluginDataDir, PluginName, zipFileName)
+	// read block meta from the remote filestore and decide if they are older than the
+	// retention period. If so, delete.
```
[the remark this answers is not in the document]
Comment (isacikgoz) [reply server/dump.go:25]: Oh this is copy pasta from sync function. Will remove.

File: server/api.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (agnivade) [reply server/api.go]: Then let's remove it. 0/5

File: server/copy.go [hunk 4]
```diff
@@ -59,3 +66,129 @@ func copyFile(srcFile, dstFile string, wr WriterFunc) error {
 
 	return nil
 }
+
+func readFromFileStore(dst, src string, b filestore.FileBackend) error {
```
Comment (agnivade) [inline server/copy.go:70]: This name still isn't changed btw. :)

**Review by agnivade — APPROVED** [review]

File: server/copy.go (hunk 4 above)
Comment (isacikgoz) [reply server/copy.go]: Right, that escaped from me :)

File: server/copy.go [hunk 5]
```diff
@@ -59,3 +66,129 @@ func copyFile(srcFile, dstFile string, wr WriterFunc) error {
 
 	return nil
 }
+
+func copyFromFileStore(dst, src string, b filestore.FileBackend) error {
+	if ok, err := b.FileExists(src); !ok {
+		return nil
```
Comment (cpoile) [inline server/copy.go:72]: Not sure if it's important, but it's possible for `FileExists` to return false with a "unable to know if file %s exists` error, but then we wouldn't return the error here -- is that ok?

**Review by cpoile — APPROVED** [review]
Nice work @isacikgoz!

File: server/copy.go (hunk 5 above)
Comment (isacikgoz) [reply server/copy.go:72]: Good point, it's actually okay as we will continue to the loop where we call the copy function.

**Pull request merged**